### PR TITLE
gh-92210: Port of Cinder's socket.__init__ which uses argument clinic.

### PR DESCRIPTION
**Pull request by zitterbewegung**
# gh-92210: Port of Cinder's socket.__init__ which uses argument clinic.

```
gh-92210: In socket library I took their code that enhanced socket.__init__ to use argument clinic to the main branch.
```

### CODE REVIEW

File: Modules/socketmodule.c [hunk 1]
```diff
@@ -5103,14 +5115,25 @@ static int sock_cloexec_works = -1;
...
 static int
-sock_initobj(PyObject *self, PyObject *args, PyObject *kwds)
+_socket_socket___init___impl(PySocketSockObject *self, int family, int type,
+                              int proto, PyObject *fdobj)
+ /*[clinic end generated code: output=8d20ec7eb58df701 input=c0966eb4587c4634]*/
 {
-    PySocketSockObject *s = (PySocketSockObject *)self;
-    PyObject *fdobj = NULL;
+    //PySocketSockObject *s = (PySocketSockObject *)self;
+    //PyObject *fdobj = NULL;
```
Comment (erlend-aasland) [inline Modules/socketmodule.c]: Please remove all dead code (git takes care of history). Applies to the whole PR.

File: Modules/socketmodule.c [hunk 2]
```diff
@@ -5103,14 +5115,25 @@ static int sock_cloexec_works = -1;
 #endif
 
 /*ARGSUSED*/
+
+/*[clinic input]
+_socket.socket.__init__
```
Comment (erlend-aasland) [inline Modules/socketmodule.c]: You can reduce the amount of lines changed with this trick:

```suggestion
_socket.socket.__init__ as sock_initobj
```

File: Modules/socketmodule.c (hunk 1 above)
Comment (zitterbewegung) [reply Modules/socketmodule.c]: okay will do.

File: Misc/NEWS.d/next/Library/2022-05-03-15-55-04.gh-issue-92210.L3xp33.rst [hunk 3]
```diff
@@ -0,0 +1 @@
+Port of Cinder's socket.__init__ which uses argument clinic cpython.
```
Comment (erlend-aasland) [inline Misc/NEWS.d/next/Library/2022-05-03-15-55-04.gh-issue-92210.L3xp33.rst]: You've now got two NEWS entries ;)

Also note that it's spelled Argument Clinic. Maybe you also want to wrap `socket.__init__` in double backticks, or maybe use a reST markup to link to the socket class.

**Comment by erlend-aasland** [comment]
Please use `git pull --no-ff main` instead of force-pushing. Force-pushing does not play well with GitHub's UI; OTOH, pulling does. It makes reviewing easier.

File: Misc/NEWS.d/next/Library/2022-05-03-17-29-04.gh-issue-92210.csDOQM.rst [hunk 4]
```diff
@@ -0,0 +1 @@
+Port of Cinder's ``socket.__init__`` which uses argument clinic cpython.
```
Comment (erlend-aasland) [inline Misc/NEWS.d/next/Library/2022-05-03-17-29-04.gh-issue-92210.csDOQM.rst]: `git rm "Misc/NEWS.d/next/Library/2022-05-03-17-29-04.gh-issue-92210.csDOQM.rst"`

Comment (erlend-aasland) [reply Misc/NEWS.d/next/Library/2022-05-03-17-29-04.gh-issue-92210.csDOQM.rst]: (There are still two NEWS entries; the newest you added is better than this one.)

File: Misc/NEWS.d/next/Library/2022-05-03-17-33-46.gh-issue-92210.csDOQM.rst [hunk 5]
```diff
@@ -0,0 +1 @@
+OPort of Cinder's ``socket.__init__`` which uses Argument Clinic to CPython.
```
Comment (erlend-aasland) [inline Misc/NEWS.d/next/Library/2022-05-03-17-33-46.gh-issue-92210.csDOQM.rst]:
```suggestion
Port ``socket.__init__`` to Argument Clinic. Patch by Cinder.
```

**Review by erlend-aasland — APPROVED** [review]

**Comment by erlend-aasland** [comment]
@ambv, can you kick the CI?

**Comment by JelleZijlstra** [comment]
I'm not ambv but I kicked it. I'll take a look now and probably merge.

**Review by JelleZijlstra — APPROVED** [review]

**Pull request merged**